pkg/river/json_walker.go

```diff
@@ -0,0 +1,409 @@
+type Field struct {
+	ID    string      `json:"id,omitempty"`
+	Key   string      `json:"key,omitempty"`
+	Label string      `json:"label,omitempty"`
+	Name  string      `json:"name,omitempty"`
+	Type  string      `json:"type,omitempty"`
+	Value interface{} `json:"value,omitempty"`
+	Body  []*Field    `json:"body,omitempty"`
+}
+
+// Health represents the health of a component.
+type Health struct {
+	State       string    `json:"state"`
+	Message     string    `json:"message"`
+	UpdatedTime time.Time `json:"updatedTime"`
+}
+
+const attr = "attr"
+
+// ConvertComponentToJSON converts a set of component information into a generic Field json representation.
+func ConvertComponentToJSON(
+	id []string,
+	args interface{},
+	exports interface{},
+	debug interface{},
+	references, referencedby []string,
+	health *Health,
+	original string,
+) (*ComponentField, error) {
+
+	nf := &ComponentField{
+		Field: Field{
+			ID:    strings.Join(id, "."),
+			Name:  strings.Join(id[0:2], "."),
+			Type:  "block",
+			Value: nil,
+		},
+		References:   references,
+		ReferencedBy: referencedby,
+		Health:       health,
+		Original:     original,
+	}
+	if len(id) == 3 {
+		nf.Label = id[2]
+	}
+
+	cArgs, err := convertComponentChild(args)
+	if err != nil {
+		return nil, err
+	}
+	if cArgs != nil {
+		nf.Arguments = cArgs
+	}
+	cExports, err := convertComponentChild(exports)
+	if err != nil {
+		return nil, err
+	}
+	if cExports != nil {
+		nf.Exports = cExports
+	}
+	cDebug, err := convertComponentChild(debug)
+	if err != nil {
+		return nil, err
+	}
+	if cDebug != nil {
+		nf.DebugInfo = cDebug
+	}
+	return nf, nil
+}
+
+// convertComponentChild is used to convert arguments, exports, health and debuginfo.
+func convertComponentChild(input interface{}) ([]*Field, error) {
+	if input == nil {
+		return nil, nil
+	}
+	_, vt, vIn := getActualValue(input)
+	fields := make([]*Field, 0)
+	rt := rivertags.Get(vt)
+	for _, t := range rt {
+		fieldValue := vIn.FieldByIndex(t.Index)
+		fieldIn, fieldT, fieldVal := getActualValue(fieldValue.Interface())
+		// Array blocks can only happen at this level
+		if t.IsBlock() && (fieldT.Kind() == reflect.Array || fieldT.Kind() == reflect.Slice) {
+			for i := 0; i < fieldVal.Len(); i++ {
+				arrEle := fieldVal.Index(i).Interface()
+				found, err := convertBlock(arrEle, &t)
+				if err != nil {
+					return nil, err
+				}
+				if found != nil {
+					fields = append(fields, found)
+				}
+			}
+		} else if t.IsBlock() {
+			found, err := convertBlock(fieldIn, &t)
+			if err != nil {
+				return nil, err
+			}
+			if found != nil {
+				fields = append(fields, found)
+			}
+		} else {
+			found, err := convertAttribute(fieldIn, vt, &t)
+			if err != nil {
+				return nil, err
+			}
+			if found != nil {
+				fields = append(fields, found)
+			}
+		}
+	}
+	return fields, nil
+}
+
+// convertThing is the switchboard for conversions
+func convertThing(in interface{}, f *rivertags.Field) (*Field, error) {
```

Comment:
Can we give this a better name? The naming is making this a bit harder to review. 

If it helps, the hierarchy is always:

- Root is a body 
- A body contains a list of attributes and blocks 
- An attribute is a name, value pair 
- A block contains a name, optional label, and inner body
- Values are either nulls, bools, numbers, strings, functions, capsules, arrays, or objects.
- Arrays contain a list of values.
- Objects contain a list of name, value pairs.